Add blogpost for 0.13.0-RC1 and 0.12.0 (#5992)
lampepfl/dotty: docs/blog/_posts/2019-03-04-13th-dotty-milestone-release.md

---
layout: blog-page

> title: Announcing Dotty 0.13.0-RC1 with Spark support and redesigned implicits

Reviewer: Worth emphasizing toplevel defs:
```suggestion
title: Announcing Dotty 0.13.0-RC1 with Spark support, top level definitions and redesigned implicits
```

Dotty is the project name for technologies that are being considered for
inclusion in Scala 3. Scala has pioneered the fusion of object-oriented and
functional programming in a typed setting. Scala 3 will be a big step towards
realising the full potential of these ideas. Its main objectives are to

- become more opinionated by promoting programming idioms we found to work well,
- simplify where possible,
- eliminate inconsistencies and surprising behaviours,
- build on strong foundations to ensure the design hangs together well,
- consolidate language constructs to improve the language’s consistency, safety, ergonomics, and
  performance.

You can learn more about Dotty on our [website](https://dotty.epfl.ch).

<!--more-->

This is our 13th scheduled release according to our
[6-week release schedule](https://dotty.epfl.ch/docs/usage/version-numbers.html).

# What’s new in the 0.13.0-RC1 technology preview?

## Experimental support for Spark

Dotty projects have always been able to [depend on Scala 2
libraries](https://github.com/lampepfl/dotty-example-project#getting-your-project-to-compile-with-dotty),
and this usually works fine as long as the Dotty code does not call a Scala 2

> macro directly. However, [Spark](http://spark.apache.org/) was known to not work

Reviewer 2: Parenthesize "`as long as the Dotty code does not call a Scala 2 macro directly`", as it is not an essential part of that sentence, and it currently confuses the point you're trying to get across.

Scala 2, and that was enough to make our Spark assignments run correctly! This
doesn't mean that our support is perfect however, so don't hesitate to [open an
issue](http://github.com/lampepfl/dotty/issues) if something is amiss.

## Introducing top level definitions

_Top level_ definitions are now supported. This means that package objects are
now redundant, and will be phased out. This means that all kinds of definitions
can be written at the top level.

```scala
package p

type Labelled[T] = (String, T)

val a: Labelled[Int] = ("count", 1)
def b = a._2
```

You can read about [dropping package
objects](https://dotty.epfl.ch/docs/reference/dropped-features/package-objects.html)
at the documentation linked or at the relevant PR
[#5754](https://github.com/lampepfl/dotty/pull/5754).

## All things impl... implied

Scala's implicits are its most distinguished feature. They are _the_ fundamental
way to abstract over context. They represent a single concept with an extremely
varied number of use cases, among them: implementing type classes, establishing
context, dependency injection, expressing capabilities, computing new types and
proving relationships between them.

> 

Reviewer 2: ```suggestion
However, with great power comes great responsibility. The current design of implicits has shown some limitations, which we have been trying to identify and address to make Scala a clearer and more pleasant language.

First of all, we found that the syntactic similarity was too great between implicit _conversions_ and implicit _values_ that depend on other implicit values, which tends to confuse newcomers. One can see that clearly in the code snippet below:
```

Author: Wrote a combination of the two.

> However, we identify a few consequences that implicits gave rise to, as a

Reviewer 2: This use of the present tense ("we identify") is rather weird, especially in conjunction with the past tense for what it is that's identified ("gave rise to"). Why not say everything in the past tense? "_We identified ...._"

> programming style. Firstly, users used implicit conversions between types, in an

Reviewer 2: ```suggestion
programming style. Firstly, some users used implicit conversions in an
```
Say "_**some**_ users", as not all of them did!

unprincipled matter. This overuse of implicit conversions decluttered code for

> sure, but it made it harder for people to reason about.

Reviewer 2: ```suggestion
While implicit conversions can be useful to remove clutter, their abuse makes it harder for people to reason about the code.
```

```scala
implicit def i1(implicit x: T): C[T] = ... // 1: conditional implicit value
implicit def i2(x: T): C[T] = ...          // 2: implicit conversion
```

The `implicit` keyword is used for both implicit conversions and conditional
implicit values and we identify that their semantic differences must be

> communicated more clearly syntactically. Secondly, implicits pose challenges for

Reviewer 2: > implicits pose challenges for tooling

What does that have to do with anything? Does the new design help in any way here?

Author: Was related to the second comment here: https://github.com/lampepfl/dotty/pull/5458

Reviewer 2: I'm not following. The second comment is by fanf who says "the benefits to help with the tools not made clear", confirming lihaoyi's question: "How does changing implicit val or implicit def for witness help tooling?", a question that was never actually answered AFAIK. There is clear no benefit to tooling from this change that I know of.

Author: Until we have more progress in this area we shouldn't mention tooling. Ok, I removed the sentence.

tooling such as error reporting for failed implicit searches. Furthermore, the

> `implicit` keyword is way too overloaded (implicit vals, defs, objects,

Reviewer 2: ```suggestion
`implicit` keyword is ascribed too many overloaded meanings in the language (implicit vals, defs, objects,
```

parameters). For instance, a newcomer can easily confuse the two

> examples above while they demonstrate completely different things, a typeclass

Reviewer 2: ```suggestion
examples above, although they demonstrate completely different things, a typeclass
```

instance is an implicit object or val if unconditional and an implicit def with
implicit parameters if conditional; arguably all of them are surprisingly
similar (syntactically). Another consideration is that the `implicit` keyword
annotates a whole parameter section instead of a single parameter, and passing
an argument to an implicit parameter looks like a regular application. This is
problematic because it can create confusion regarding what parameter gets passed
in a call. Last but not least, sometimes implicit parameters are merely

> propagated in nested function calls and not used at all, so names of implicit

Reviewer 2: ```suggestion
propagated in nested function calls and not used at all, so giving names to implicit parameters is often redundant and only adds noise to a function signature.
```

}

implied ListOrd[T] given (ord: Ord[T]) for Ord[List[T]] {
  def compare(xs: List[T], ys: List[T]): Int = (xs, ys) match {
    case (Nil, Nil) => 0
    case (Nil, _) => -1
    case (_, Nil) => +1
    case (x :: xs1, y :: ys1) =>
      val fst = ord.compare(x, y)
      if (fst != 0) fst else xs1.compareTo(ys1)
  }
}
```

A `given` clause can also designate an inferable parameter for functions:

```scala
def max[T](x: T, y: T) given (ord: Ord[T]): T =
  if (ord.compare(x, y) < 1) y else x
```

With this scheme all invocations of the `max` function below are equally valid:
```scala
max(2, 3) given IntOrd
max(List(1, 2, 3), Nil)
max(2, 3)
```

We introduce _Anonymous Implied Instances_ which are used when we do not need a name for an implied
instance:

```scala
implied for Ord[Int] { ... }
```

For convenience, we also introduce _Implied Alias Instances_. They offer aliases
for implied instances. For example, the line below offers an alias with the name
`ctx` (could also be anonymous if name can be omitted). Each time an implied
instance of `ExecutionContext` is demanded the right-hand side is returned.

```scala
implied ctx for ExecutionContext = currentThreadPool().context
```

> For symmetry, we define our well-known `implicitly` from `Predef` in terms of

Reviewer 2: Does it really matter how `implicitly` is implemented?

```suggestion
We have also added a synonym to `implicitly`, which is often more natural to spell out in user code
```

(Apparently we'd keep both `the` and `implicitly` https://github.com/lampepfl/dotty/pull/5893#issuecomment-463634860 )

**You can read more about** the alternative to implicits through the *Contextual
Abstractions* section of our documentation or for a deep dive from the relevant
PR chain that originated from
[#5458](https://github.com/lampepfl/dotty/pull/5458). The syntax changes for new
implicits are summarized in
[#5825](https://github.com/lampepfl/dotty/pull/5825).

This release offers the support for _typeclass derivation_ as a language
feature. Typeclass derivation is a way to generate instances of certain type
classes automatically or with minimal code hints, and is now supported natively
with *dedicated language support*. A type class in this sense is any trait or
class with a type parameter that describes the type being operated on. Commonly
used examples are `Ordering`, `Show`, or `Pickling`. We introduce a new
`derives` clause that generates implied instances of the `Eql`, `Ordering`, and
`Pickling` traits in the companion object `Tree`. Take a look at the example
below:

```scala
enum Tree[T] derives Eql, Ordering, Pickling {
  case Branch(left: Tree[T], right: Tree[T])
  case Leaf(elem: T)
}
```

where the generated implied instances are the ones below:
```scala

> implied [T: Eq]       for Eq[Tree[T]]       = Eq.derived

Reviewer 3: Eql instead of Eq

```scala
def derived[T] given Generic[T] = ...
```

You can read more about [Typeclass
Derivation](https://dotty.epfl.ch/docs/reference/contextual/derivation.html) or
for a deep dive at the relevant PRs:
[#5540](https://github.com/lampepfl/dotty/pull/5540) and
[#5839](https://github.com/lampepfl/dotty/pull/5839).

_Multiversal equality_ is now supported through the `Eql` marker trait (renamed
from `Eq` to differentiate it from Cats' `Eq`). For example, in order to be able
to compare integers with strings now, instead of a custom implicit we can
provide a derived implicit instance:

```scala
implied for Eql[Int, String] = Eql.derived
```

You can read how we based multiversal equality on typeclass derivation through
the relevant PR [#5843](https://github.com/lampepfl/dotty/pull/5843).

_Implicit conversions_ are now defined by implied instances of the
`scala.Conversion` class. For example:

```scala
implied for Conversion[String, Token] {
  def apply(str: String): Token = new KeyWord(str)
}
```

> _Top level_ definitions are now supported. This means that package objects are

Reviewer: This should be a separate section with its own heading `## Introducing top level definitions`, and I'd move it above the long description of implicits.

Author: I will put on on a separate heading but I'll leave it after implied (ordered-by importance 🥇)

Reviewer: > ordered-by importance

That's reasonable, my reasoning was that we should put the short and easy-to-understand stuff first to make sure people actually read it, otherwise I'm afraid the implied stuff will eclipse everything else. One way to avoid that would be to move the long description of the implied stuff in the documentation and link it from the blog post.

Reviewer 4: I agree with @smarter. Put the short stuff first. Also it's a non-controversial, super-positive change that will bring some feel-good sentiment.

Author: The implied section is just a medley of all the various elements from the documentation so it will not add any value to it. The remaining two (implicit resolution changes and SemanticDB generator) cannot be on top of implied so I propose lets keep this order. However, I improved the very first sentence to reflect the updated title 😉

Author: Oh well, you won! 😆I am moving the top level after spark.

Author: The 

```scala
implied {
  def (x: C) pair (y: C) = (x, y)
}
```

has to go now, as it is a spoiler alert for the section to come.

Reviewer 5: It's also syntax that isn't introduced.  (What even is it?  An implied block of definitions?  Can't get it to do anything useful in the REPL)

Author: https://dotty.epfl.ch/docs/reference/contextual/relationship-implicits.html#anonymous-implied-instances

now redundant, and will be phased out. This means that all kinds of definitions
can be written at the top level.

```scala

> package p.

Reviewer: ```suggestion
package p
```

type Labelled[T] = (String, T)
val a: Labelled[Int] = ("count", 1)
def b = a._2

> case class C()

Reviewer 5: ```suggestion
final case class C()
```

A good practice, IMO.

> implicit object Cops {

Reviewer: Might be confusing to mention this after talking at length about the new proposed syntax for implicits...

  def (x: C) pair (y: C) = (x, y)
}
```

You can read about [dropping package
objects](https://dotty.epfl.ch/docs/reference/dropped-features/package-objects.html)
at the documentation linked or at the relevant PR
[#5754](https://github.com/lampepfl/dotty/pull/5754).

> **This blogpost offers only a brief summary of the new features, for more details

Reviewer: This is about implicits, not toplevel defs, needs to be moved.
